examples/mixed_signal/tests/test_rescap.py

```diff
@@ -0,0 +1,168 @@
+# Copyright cocotb contributors
+# Licensed under the Revised BSD License, see LICENSE for details.
+# SPDX-License-Identifier: BSD-3-Clause
+
+from cocotb.regression import TestFactory
+from cocotb.triggers import Timer, ReadOnly
+from cocotb.utils import get_sim_time
+
+from collections import namedtuple, defaultdict
+
+from itertools import cycle
+
+import matplotlib.pyplot as plt
+
+Dataset = namedtuple("Dataset", "time, voltage, current")
+
+
+class ResCap_TB(object):
+    """The testbench class for the rescap design."""
+    def __init__(self, tb_hdl):
+        self.tb_hdl = tb_hdl
+        self.analog_probe = (
+            tb_hdl.i_analog_probe
+        )  #: The instance name of the analog probe module.
+        self.togglestream = cycle(range(2))  # toggle between 0 and 1
+
+    async def _get_single_sample(self, node):
+        toggle = next(self.togglestream)
+        self.tb_hdl.i_analog_probe.node_to_probe <= node.encode('ascii')
+        self.analog_probe.probe_voltage_toggle <= toggle
+        self.analog_probe.probe_current_toggle <= toggle
+        await ReadOnly()
+        dataset = Dataset(
+            time=get_sim_time(units="ns"),
+            voltage=self.analog_probe.voltage.value,
+            current=self.analog_probe.current.value * 1000.0  # in mA
+        )
+        self.tb_hdl._log.debug(
+            "{}={:.4} V, {:.4} mA".format(
+                self.analog_probe.node_to_probe.value.decode("ascii"), dataset.voltage, dataset.current
+            )
+        )
+        return dataset
+
+    async def get_sample_data(self, nodes, num=1, delay_ns=1):
+        """For all *nodes*, get *num* samples, spaced *delay_ns* apart.
+
+        Yields:
+            list: List (*num* samples long) of :any:`Dataset` for all *nodes*.
+        """
+        if not isinstance(nodes, list):  # single element? make it a list
+            _nodes = [nodes]
+        else:
+            _nodes = nodes
+        datasets = defaultdict(list)
+        for idx in range(num):
+            for node in _nodes:
+                dataset = await self._get_single_sample(node)
+                datasets[node].append(dataset)
+                if num > 1:
+                    await Timer(delay_ns, units="ns")
```

Comment:
`get_single_sample` passes time. I guess you need to do that for the probe to function?
The `Timer` seem like it should be dedented so you aren't waiting additional time between sampling each node.
Finally, `if num > 1`, on the last sample, it's going to wait. It should probably be `if idx != num`.

Reply:
Added a comment that a small waiting time is needed for the analog values to update.
Timer was wrongly indented, thanks.
About the last sample, it's a matter of taste I guess... The way it is now you could call this function several times back to back and maintain equal spacing. OTOH, we cannot go back in time once we have spent it, so not ending with a waiting time is more general.